File: 1105-filling-bookcase-shelves/1105-filling-bookcase-shelves.cs
public class Solution {
    int[,] dp;

    public int MinHeightShelves(int[][] books, int shelfWidth) {
        dp = new int[books.Length, shelfWidth+1];
        return MinHeightShelves(books, shelfWidth, 0, shelfWidth, 0);
    }

    private int MinHeightShelves(int[][] books, int shelfWidth, int idx, int width, int height){
        if(idx == books.Length) return 0;
        if(dp[idx, width] != 0) return dp[idx, width];

        var (bookWidth, bookHeight) = (books[idx][0], books[idx][1]);
        var currHeight = bookHeight + MinHeightShelves(books, shelfWidth, idx+1, shelfWidth - bookWidth, bookHeight);
        if(bookWidth <= width){
            var heightIncr = bookHeight > height ? bookHeight - height : 0;
            var sameRack = heightIncr + MinHeightShelves(books, shelfWidth, idx+1, width - bookWidth, height + heightIncr);
            currHeight = Math.Min(currHeight, sameRack);
        }

        dp[idx, width] = currHeight;
        return currHeight;
    }
}

File: 1105-filling-bookcase-shelves/1105-filling-bookcase-shelves.java
class Solution {
    int[][] dp;

    public int minHeightShelves(int[][] books, int shelfWidth) {
        dp = new int[books.length][shelfWidth+1];
        return minHeightShelves(books, shelfWidth, 0, shelfWidth, 0);
    }

    private int minHeightShelves(int[][] books, int shelfWidth, int idx, int width, int height){
        if(idx == books.length) return 0;
        if(dp[idx][width] != 0) return dp[idx][width];

        int bookWidth = books[idx][0], bookHeight = books[idx][1];
        var currHeight = bookHeight + minHeightShelves(books, shelfWidth, idx+1, shelfWidth - bookWidth, bookHeight);
        if(bookWidth <= width){
            var heightIncr = Math.max(0, bookHeight - height);
            var sameRack = heightIncr + minHeightShelves(books, shelfWidth, idx+1, width - bookWidth, height + heightIncr);
            currHeight = Math.min(currHeight, sameRack);
        }

        dp[idx][width] = currHeight;
        return currHeight;
    }
}